File: e2e_tests/python/chat_session.py
import asyncio
import json
import logging
from typing import Optional

from server_clients.server import Server
from server_clients.servers import Servers
from llm_client import LLMClient


class ChatSession:
    """Orchestrates the interaction between user, LLM, and tools."""

    def __init__(
        self, servers: list[Server], llm_client: LLMClient, user_utterances: list[str]
    ) -> None:
        self.servers: list[Server] = servers
        self.llm_client: LLMClient = llm_client
        self.user_utterances: list[str] = user_utterances

    async def execute_requested_tools(
        self, servers_manager: Servers, llm_response
    ) -> Optional[list[dict]]:
        """Process the LLM response and execute tools if needed.

        Args:
            llm_response: The response from the Bedrock Converse API.

        Returns:
            The result of tool execution, if any.
        """
        stop_reason = llm_response["stopReason"]

        if stop_reason == "tool_use":
            try:
                tool_responses = []
                for content_item in llm_response["output"]["message"]["content"]:
                    if "toolUse" in content_item:
                        logging.info(
                            f'Executing tool: {content_item["toolUse"]["name"]}'
                        )
                        logging.info(
                            f'With arguments: {content_item["toolUse"]["input"]}'
                        )
                        response = await servers_manager.execute_tool(
                            content_item["toolUse"]["name"],
                            content_item["toolUse"]["toolUseId"],
                            content_item["toolUse"]["input"],
                        )
                        tool_responses.append(response)
                return {"role": "user", "content": tool_responses}
            except KeyError as e:
                raise ValueError(f"Missing required tool use field: {e}")
            except Exception as e:
                raise ValueError(f"Failed to execute tool: {e}")
        else:
            # Assume this catches stop reasons "end_turn", "stop_sequence", and "max_tokens"
            return None

    async def start(self) -> None:
        """Main chat session handler."""
        async with Servers(self.servers) as server_manager:
            all_tools = await server_manager.list_tools()
            tools_description = [tool.format_for_llm() for tool in all_tools]

            system_prompt = "You are a helpful assistant."

            messages = []

            for i, user_input in enumerate(self.user_utterances):
                if i != 0:
                    print("\n**Pausing 5 seconds to avoid Bedrock throttling**")
                    await asyncio.sleep(5)

                print(f"\nYou: {user_input}")

                messages.append({"role": "user", "content": [{"text": user_input}]})

                llm_response = self.llm_client.get_response(
                    messages, system_prompt, tools_description
                )
                logging.debug("\nAssistant: %s", json.dumps(llm_response, indent=2))
                print(
                    f'\nAssistant: {llm_response["output"]["message"]["content"][0]["text"]}'
                )
                messages.append(llm_response["output"]["message"])

                tool_results = await self.execute_requested_tools(
                    server_manager, llm_response
                )

                if tool_results:
                    logging.debug(
                        "\nTool Results: %s", json.dumps(tool_results, indent=2)
                    )
                    messages.append(tool_results)
                    final_response = self.llm_client.get_response(
                        messages, system_prompt, tools_description
                    )
                    logging.debug(
                        "\nFinal response: %s", json.dumps(final_response, indent=2)
                    )
                    print(
                        f'\nAssistant: {final_response["output"]["message"]["content"][0]["text"]}'
                    )
                    messages.append(final_response["output"]["message"])

File: e2e_tests/typescript/src/chat_session.ts
import { Server } from "./server_clients/server.js";
import { Servers } from "./server_clients/servers.js";
import { LLMClient } from "./llm_client.js";
import logger from "./logger.js";
import {
  ContentBlock,
  ConverseCommandOutput,
  Message,
} from "@aws-sdk/client-bedrock-runtime";

/**
 * Orchestrates the interaction between user, LLM, and tools.
 */
export class ChatSession {
  private servers: Server[];
  private llmClient: LLMClient;
  private userUtterances: string[] = [];

  constructor(
    servers: Server[],
    llmClient: LLMClient,
    userUtterances: string[]
  ) {
    this.servers = servers;
    this.llmClient = llmClient;
    this.userUtterances = userUtterances;
  }

  /**
   * Process the LLM response and execute tools if needed.
   * @param serversManager The servers manager.
   * @param llmResponse The response from the Bedrock Converse API.
   * @returns The result of tool execution, if any.
   */
  async executeRequestedTools(
    serversManager: Servers,
    llmResponse: ConverseCommandOutput
  ): Promise<Message | null> {
    const stopReason = llmResponse.stopReason;

    if (stopReason === "tool_use") {
      try {
        const toolResponses: ContentBlock[] = [];
        for (const contentItem of llmResponse.output!.message!.content!) {
          if ("toolUse" in contentItem) {
            logger.info(`Executing tool: ${contentItem.toolUse!.name}`);
            logger.info(
              `With arguments: ${JSON.stringify(contentItem.toolUse!.input)}`
            );
            const response = await serversManager.executeTool(
              contentItem.toolUse!.name!,
              contentItem.toolUse!.toolUseId!,
              contentItem.toolUse!.input! as Record<string, any>
            );
            toolResponses.push(response);
          }
        }
        return { role: "user", content: toolResponses };
      } catch (e) {
        throw new Error(`Failed to execute tool: ${e}`);
      }
    } else {
      // Assume this catches stop reasons "end_turn", "stop_sequence", and "max_tokens"
      return null;
    }
  }

  /**
   * Main chat session handler.
   */
  async start(): Promise<void> {
    const serverManager = new Servers(this.servers);

    try {
      await serverManager.initialize();

      const allTools = await serverManager.listTools();
      const toolsDescription = allTools.map((tool) => tool.formatForLLM());

      const systemPrompt = "You are a helpful assistant.";

      const messages: Message[] = [];

      for (const [i, userInput] of this.userUtterances.entries()) {
        if (i != 0) {
          console.log("\n**Pausing 30 seconds to avoid Bedrock throttling**");
          await new Promise((resolve) => setTimeout(resolve, 30 * 1000));
        }

        console.log(`\You: ${userInput}`);

        messages.push({ role: "user", content: [{ text: userInput }] });

        const llmResponse = await this.llmClient.getResponse(
          messages,
          systemPrompt,
          toolsDescription
        );

        logger.debug("\nAssistant: " + JSON.stringify(llmResponse, null, 2));
        console.log(
          `\nAssistant: ${llmResponse.output!.message!.content![0].text}`
        );
        messages.push(llmResponse.output!.message!);

        const toolResults = await this.executeRequestedTools(
          serverManager,
          llmResponse
        );

        if (toolResults) {
          logger.debug(
            "\nTool Results: " + JSON.stringify(toolResults, null, 2)
          );
          messages.push(toolResults);
          const finalResponse = await this.llmClient.getResponse(
            messages,
            systemPrompt,
            toolsDescription
          );
          logger.debug(
            "\nFinal response: " + JSON.stringify(finalResponse, null, 2)
          );
          console.log(
            `\nAssistant: ${finalResponse.output!.message!.content![0].text}`
          );
          messages.push(finalResponse.output!.message!);
        }
      }
    } finally {
      await serverManager.close();
    }
  }
}
